lib/include/tick/array/sparsearray2d.h

```diff
@@ -107,26 +107,41 @@ class SparseArray2d : public BaseArray2d<T> {
     ar(this->_size);
 
     inner_save(ar);
+  }
 
-    std::vector<INDICE_TYPE> v_indices(this->_indices, this->_indices + this->_size_sparse),
-        v_row_indices(this->_row_indices, this->_row_indices + (this->_n_rows + 1));
-    ar(v_indices);
-    ar(v_row_indices);
+  template <class Archive, typename Y = T>
+  typename std::enable_if<std::is_same<Y, std::atomic<K>>::value
+    && cereal::traits::is_output_serializable<cereal::BinaryData<T>, Archive>::value, void>::type
+  inner_save(Archive &ar) const {
+    for (size_t i = 0; i < this->_size_sparse; i++) ar(this->_data[i].load());
+    ar(cereal::binary_data(this->_indices, sizeof(INDICE_TYPE) * this->_size_sparse));
+    ar(cereal::binary_data(this->_row_indices, sizeof(INDICE_TYPE) * (this->_n_rows + 1)));
   }
 
   template <class Archive, typename Y = T>
-  typename std::enable_if<std::is_same<Y, std::atomic<K>>::value>::type inner_save(
-      Archive &ar) const {
-    std::vector<K> k_data;
-    for (size_t i = 0; i < this->_size_sparse; i++) k_data.emplace_back(this->_data[i]);
-    ar(k_data);
+  typename std::enable_if<std::is_same<Y, std::atomic<K>>::value
```

Comment:
How did you end up knowing we need to differenciate binary and non binary cases?

Reply:
this was already the case, which came about once pickle was brought into the equation

pickling uses "JSON" and doesn't support "binary_data" 

I'd like to see if we can decouple JSON stuff from pickling too

Reply:
Ok! If we can't get rid of this, maybe we should just add it in a comment :)